apps/service_providers/llm_service/history_managers.py

```diff
@@ -0,0 +1,239 @@
+        self.node_id = node_id
+        self.history_type = history_type
+        self.history_name = history_name
+        self.max_token_limit = max_token_limit
+        self.chat_model = chat_model
+        self.trace_service = session.experiment.trace_service if session else None
+
+        self.input_message_metadata = None
+        self.output_message_metadata = None
+
+    @classmethod
+    def for_llm_chat(
+        cls,
+        session: ExperimentSession,
+        node_id: str,
+        history_type: PipelineChatHistoryTypes,
+        history_name: str,
+        max_token_limit: int,
+        chat_model: BaseChatModel,
+    ) -> Self:
+        return cls(
+            session=session,
+            node_id=node_id,
+            history_type=history_type,
+            history_name=history_name,
+            max_token_limit=max_token_limit,
+            chat_model=chat_model,
+        )
+
+    @classmethod
+    def for_assistant(cls) -> Self:
+        return cls()
+
+    def get_chat_history(self, input_messages: list):
+        if self.history_type == PipelineChatHistoryTypes.NONE or self.session is None:
+            return []
+
+        if self.history_type == PipelineChatHistoryTypes.GLOBAL:
+            return compress_chat_history(
+                chat=self.session.chat,
+                llm=self.chat_model,
+                max_token_limit=self.max_token_limit,
+                input_messages=input_messages,
+            )
+
+        try:
+            history: PipelineChatHistory = self.session.pipeline_chat_history.get(
+                type=self.history_type, name=self._get_history_name(self.node_id)
+            )
+        except PipelineChatHistory.DoesNotExist:
+            return []
+        return compress_pipeline_chat_history(
+            pipeline_chat_history=history,
+            max_token_limit=self.max_token_limit,
+            llm=self.chat_model,
+            input_messages=input_messages,
+        )
+
+    def _get_history_name(self, node_id):
+        if self.history_type == PipelineChatHistoryTypes.NAMED:
+            return self.history_name
+        return node_id
+
+    def add_messages_to_history(
+        self, input: str, input_message_metadata: dict, output: str, output_message_metadata: dict, *args, **kwargs
+    ):
+        self.input_message_metadata = input_message_metadata
+        self.output_message_metadata = output_message_metadata
+
+        if self.history_type == PipelineChatHistoryTypes.NONE:
+            return
+
+        if self.history_type == PipelineChatHistoryTypes.GLOBAL:
+            # Global History is saved outside of the node
+            return
+
+        history, _ = self.session.pipeline_chat_history.get_or_create(
+            type=self.history_type, name=self._get_history_name(self.node_id)
+        )
+
+        message = history.messages.create(human_message=input, ai_message=output, node_id=self.node_id)
+        # TODO: Save normal session history here as well
```

Comment:
is this TODO still necessary?

Reply:
The answer is..maybe. Deffs not for this PR. I think it would be great if we can move the creation of `ChatMessage` records that's currently happening outside of pipelines to in here. If you don't have anything against this plan, I'll remove the TODO and create a ticket for it

Reply:
You are correct @snopoke. I didn't think this through thoroughly. @coderabbitai please close issue #975 with the reason that simon mentioned

Reply:
@coderabbitai I confirm